torch/distributed/device_mesh.py

```diff
@@ -488,6 +325,162 @@ def __init__(
+
+            # If the layout has already been mapped to a process group backend, we ignore the
+            # backend_override and group, and return early. We also throw a warning for that.
+            if layout in layouts_to_groups_map:
+                # TODO: We need to bookkeep pg_option together with the group name.
+                if backend_override != (None, None):
+                    warnings.warn(
+                        f"Group for {layout} ({name=}) already exists, ignoring backend override"
+                    )
+                if group is not None:
+                    warnings.warn(
+                        f"Group for {layout} ({name=}) already exists, ignoring explicit group"
+                    )
+                return
+
+            # When user explicitly pass in a process group, we directly reuse that PG as backend rather
+            # than creating a new one.
+            if group is not None:
+                layouts_to_groups_map[layout] = group.group_name
+                return
+
+            default_group = _get_default_group()
+
+            if layout == _Layout(((get_world_size(), 1),)) and backend_override == (
+                None,
+                None,
+            ):
+                # Append the default pg to the first dim groups only if the default pg is compatible with `self.device_type`.
+                # Otherwise, create new pg.
+                ranks = list(range(get_world_size()))
+                group = (
+                    new_group(
+                        backend="cpu:gloo,cuda:nccl",
+                        ranks=ranks,
+                        group_desc="mesh_default",
+                    )
+                    if torch.cuda.is_available()
+                    and get_backend(default_group) == "gloo"
+                    else default_group
+                )
+            else:
+                # Generate the pg_ranks_by_dim for pg_creation from the layout.
+                pg_ranks_by_dim = array_mesh if array_mesh is not None else layout.layout_to_global_ranks(get_world_size())
+                backend, pg_options = backend_override
+
+                # If we have a 2D mesh with mesh_dim_names ("dp", "tp"), the group description
+                # of the subgroups would be `mesh_dim_dp` and `mesh_name_tp`.
+                # If the mesh doesn't not have a mesh_dim_names, then the group description of the
+                # subgroup would be `mesh_dim_0` and `mesh_dim_1`.
+                group_desc = f"mesh_{name}" if name is not None else f"mesh_dim_{dim}"
+
+                # If bound_device_id exists, it means the nccl communicator has been eagerly initialized
+                # so that we can use `split_group` to create subgroups through `ncclCommSplit`.
+                # In this case, we only need to make one API call (`split_group``) for the subgroup creation
+                # for each mesh dimension. In a 2 * 4 mesh, we only need to make 2 API calls per ranks to create
```

Comment:
where is the for loop that does these 2 or 6 api calls?

Reply:
Calling this function has a for loop that is the for-loop for 2 and in the else there is another for-loop... I just blindly copied the original comments.. I think the comment here has wrong math. let me change it.